File: treeLevels.py
# T test cases
# in each one, given a tree, 
#   find the abs difference of sums of the nodes values
#   at odd and even level

from queue import Queue

def printQueue(q):
    print("Queue:", end=' ')
    q2 = Queue()
    while q.qsize():
        x = q.get()
        print(x, end=' ')
        q2.put(x)
        print()
    while q2.qsize():
        q.put(q2.get())

def printSets(even, odd):
    print(f"Odd: {odd}")
    print(f"Even: {even}")

def addEdge(d, a, b):
    d[a][1].append(b)
    d[b][1].append(a)
    #print(d)

def treeInit(d, N):
    nodesValues = list(map(lambda x: int(x), list(input().split(' '))))
    for i in range(N):
        d[i]=[nodesValues[i], []]
    for i in range(N-1):
        edge = list(map(lambda x: int(x), list(input().split(' '))))
        addEdge(d, edge[0]-1, edge[1]-1)
    #print(d)

def BFS(d, even, odd):
    q = Queue()
    seen = set()
    q.put((1, 1))
    #printQueue(q)
    odd.add(1)
    #printSets(even, odd)
    while q.qsize():
        currL = q.get()
        #print(f"I popped:{currL}")
        par = currL[1]
        curr = currL[0]
        if curr not in seen:
            seen.add(curr)
            if par!=curr:
                if par in odd:
                    even.add(curr)
                else:
                    odd.add(curr)
            #printSets(even, odd)
            for node in d[curr][1]:
                q.put((node, curr))
                #print(f"I put ({node}, {curr}) in the queue")
        
def diffCompute(tree, even, odd):
    sum1 = 0
    sum2 = 0
    for v in even:
        sum1+=tree[v][0]
    for u in odd:
        sum2+=tree[u][0]
    print(abs(sum1-sum2))

def solveOne():
    N = int(input())
    tree = {}
    even = set()
    odd=set()
    treeInit(tree, N)
    BFS(tree, even, odd)
    diffCompute(tree, even, odd)

def solveAll():
    T = int(input())
    for i in range(T):
        solveOne()

solveAll()

File: treeLevels.cpp
/*
 T test cases
 in each one, given a tree, 
   find the abs difference of the sums of the nodes values
   at odd and even level
*/

#include <iostream>
using namespace std;
#include <set>
#include <vector>
#include <queue>
#include <tuple>
vector<vector<int>> tree;
vector<bool> seen;
vector<bool> odd;
int N;

void printVec(vector<bool> vec) {
    for (int i=0; i<int(vec.size()); i++) {
        cout << vec[i] << " ";
    }
    cout << endl;
}

void printTree() {
    cout << "Tree:" << endl;
    for (int i=0; i<int(tree.size()); i++) {
        cout << i << ": ";
        for (int j=0; j<int(tree[i].size()); j++) {
            cout << tree[i][j] << ", ";
        }
        cout << endl;
    }
}

void addEdge(int a, int b) {
    tree[a].push_back(b);
    tree[b].push_back(a);
}

void treeInit() {
    for (int i=0; i<N; i++) {
        int x;
        cin >> x;
        vector<int> line;
        line.push_back(x);
        tree.push_back(line);
    }
    //printTree();
    for (int i=0; i<N-1; i++) {
        int a, b;
        cin >> a;
        cin >> b;
        addEdge(a-1, b-1);
    }
    //printTree();
}

void BFS() {
    queue<tuple<int, int>> q;
    q.push(tuple<int, int>{0,0});
    odd[0]=true;
    while (!q.empty()) {
        tuple<int, int> t = q.front();
        q.pop();
        int curr = get<0>(t);
        int par = get<1>(t);
        //cout << "Popped from queue: (" << curr << ", " << par <<")" << endl;
        if (!seen[curr]) {
            seen[curr] = true;
            if (par!=curr) {
                if (!odd[par]) {
                    odd[curr]=true;
                }
            }
            for (int i=1; i<int(tree[curr].size()); i++) {
                q.push(tuple<int, int>{tree[curr][i], curr});
            }
        }
    }    
}

void diffCompute() {
    int sum1 = 0;
    int sum2 = 0;
    for (int i=0; i<int(odd.size()); i++) {
        if (odd[i]) {
            sum1 += tree[i][0]; 
        }
        else {
            sum2 += tree[i][0];
        }
    }
    cout << (sum1>sum2 ? sum1-sum2 : sum2-sum1) << endl;;
}

void init() {
    cin >> N;
    //cout << "N: " << N << endl;
    tree.clear();
    seen.clear();
    odd.clear();
    for (int i=0; i<N; i++) {
        seen.push_back(false);
        odd.push_back(false);
    }
}

void solveOne() {
    init();
    treeInit();
    BFS();
    diffCompute();
}

int main() {
    int T;
    cin >> T;
    //cout <<"T: " << T << endl;
    for (int i=0 ; i<T; i++) {
        solveOne();
    }
}